src/core/meshcop/tcat_agent.cpp

```diff
@@ -500,6 +503,66 @@ Error TcatAgent::HandleStartThreadInterface(void)
     return error;
 }
 
+uint8_t *TcatAgent::GetAdvertisementData(uint16_t &aLen)
+{
+    Message *buffer = Get<MessagePool>().Allocate(Message::kTypeBle, 0);
+
+    if (buffer == nullptr || mVendorInfo == nullptr)
+    {
+        aLen = 0;
+        return nullptr;
+    }
+
+    *reinterpret_cast<uint16_t *>(mAdvertisement) = LittleEndian::HostSwap16(OT_TOBLE_SERVICE_UUID);
+    mAdvertisement[2]                             = OT_TCAT_VERSION << 4 | OT_TCAT_OPCODE;
+
+    IgnoreError(buffer->AppendBytes(mAdvertisement, 3));
+
+    switch (MapEnum(mVendorInfo->mDeviceIdType))
+    {
+    case kTcatDeviceIdOui24:
+        IgnoreError(ot::Tlv::Append<VendorOuiTlv24>(*buffer, mVendorInfo->mDeviceId));
+        break;
+    case kTcatDeviceIdOui36:
+        IgnoreError(ot::Tlv::Append<VendorOuiTlv36>(*buffer, mVendorInfo->mDeviceId));
+        break;
+    case kTcatDeviceIdDiscriminator:
+        IgnoreError(ot::Tlv::Append<DeviceDiscriminatorTlv>(*buffer, mVendorInfo->mDeviceId));
+        break;
+    case kTcatDeviceIdIanaPen:
+        IgnoreError(ot::Tlv::Append<VendorIanaPenTlv>(*buffer, mVendorInfo->mDeviceId));
+        break;
+    default:
+        break;
+    }
+
+    otBleLinkCapabilities caps = otPlatGetBleLinkCapabilities(&GetInstance());
+
+    IgnoreError(ot::Tlv::Append<BleLinkCapabilitiesTlv>(*buffer, *reinterpret_cast<uint8_t *>(&caps)));
+
+    DeviceTypeAndStatus tas;
+    tas.mRsv                 = 0;
+    tas.mIsCommisionned      = Get<ActiveDatasetManager>().IsCommissioned();
+    tas.mThreadNetworkActive = Get<Mle::Mle>().IsAttached();
+    tas.mDeviceType          = Get<Mle::Mle>().GetDeviceMode().IsFullThreadDevice();
+    tas.mRxOnWhenIdle        = Get<Mle::Mle>().GetDeviceMode().IsRxOnWhenIdle();
+
+#if OPENTHREAD_FTD && (OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE || OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE)
+    tas.mIsBorderRouter = true;
+#else
+    tas.mIsBorderRouter = false;
+#endif
+
+    IgnoreError(ot::Tlv::Append<DeviceTypeAndStatusTlv>(*buffer, *reinterpret_cast<uint8_t *>(&tas)));
+
+    aLen = buffer->GetLength();
+    OT_ASSERT(mAdvertisementLength <= OT_TCAT_ADVERTISEMENT_MAX_LEN);
```

Comment:
with this assert, it can be triggered if the user types a CLI command `tcat vendorid`. That seems not robust enough. Maybe an error can be logged? And the message truncated?

Reply:
OT_ASSER will print a message. The assert is there to guard against not setting too big advertisement data.